BUG: Set check_exact to true if dtype is int

parthi-siva opened the pull request:
- [x] closes #55882 
- [x] [Tests added and passed]
- [x] All [code checks passed]
- [ ] Added [type annotations]
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

parthi-siva committed:
BUG: Set check_exact to true if dtype is int

parthi-siva commented:
@MarcoGorelli not sure where to add test cases for the change.

parthi-siva committed:
BUG: Add check to ignore dytype difference

- If int dtype is different we are ignoring the difference
- so added check to set check_exact to true only when dtype is same

parthi-siva committed:
TST: Added test cases

parthi-siva committed:
TST: Fix failing test cases

MarcoGorelli commented on pandas/_testing/asserters.py:
```diff
@@ -921,6 +921,14 @@ def assert_series_equal(
+    if (
+        is_integer_dtype(left.dtype)
+        and is_integer_dtype(right.dtype)
+        and isinstance(left._values, type(right._values))
+        and isinstance(right._values, type(left._values))
```
just for my understanding, why do we need these two lines?

parthi-siva replied:
Sure.
 In this test case pandas/tests/series/test_constructors.py::TestSeriesConstructors::test_constructor_maskedarray:572
one assertion is failing.

```
    def test_constructor_maskedarray(self):
        data = ma.masked_all((3,), dtype=float)
        result = Series(data)
        expected = Series([np.nan, np.nan, np.nan])
        tm.assert_series_equal(result, expected)
    
        data[0] = 0.0
        data[2] = 2.0
        index = ["a", "b", "c"]
        result = Series(data, index=index)
        expected = Series([0.0, np.nan, 2.0], index=index)
        tm.assert_series_equal(result, expected)
    
        data[1] = 1.0
        result = Series(data, index=index)
        expected = Series([0.0, 1.0, 2.0], index=index)
        tm.assert_series_equal(result, expected)
    
        data = ma.masked_all((3,), dtype=int)
        result = Series(data)
        expected = Series([np.nan, np.nan, np.nan], dtype=float)
        tm.assert_series_equal(result, expected)
    
        data[0] = 0
        data[2] = 2
        index = ["a", "b", "c"]
        result = Series(data, index=index)
        expected = Series([0, np.nan, 2], index=index, dtype=float)
        tm.assert_series_equal(result, expected)
    
        data[1] = 1
        result = Series(data, index=index)
        expected = Series([0, 1, 2], index=index, dtype=int)
>       tm.assert_series_equal(result, expected)
E       AssertionError: Series are different
E       
E       Series classes are different
E       [left]:  MaskedArray
E       [right]: ndarray
```

Even though `left.dtype `and` right.dtype `are same their type are different. So here are setting check_exact to true only when the types of left and right are same.

parthi-siva committed:
DOC: Update function documentation

parthi-siva commented:
> Well I'm relieved that CI is still green 😄 Can you update the docs for this method to note that `rtol`, `atol` and `check_exact` don't take effect for int dtype?
> 
> Regarding the check - not totally sure about the check, will take a closer look next week

Sure. Updated the documentation. Added note under `check_exact`. Since we are setting it to True in case of int dtype. Also, documentation for `rtol` and `atol` states it is applicable only when check_exact is False I assumed that it is sufficient to add not under `check_exact`

MarcoGorelli committed:
check_exact only takes effect for floating dtypes

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/main' into BUG-GH#55882

MarcoGorelli commented:
I think it should always be exact, except for floats?

You should also check for extensions arrays, and include a whatsnew

I've added a commit anyway, hope it's ok - do the changes look alright to you?

MarcoGorelli committed:
xfail failing test

parthi-siva commented:
Yeah, make sense.

Some testcases are failing in some builds. Is that alright?

MarcoGorelli commented:
Looks like another bug was hiding because of the check_exact default 😄  https://github.com/pandas-dev/pandas/issues/56136

For now let's xfail the relevant test

MarcoGorelli reviewed (approved):
Looks good to me - although I did add a commit, so leaving open a bit in case others have objections

parthi-siva commented:
Sure Marco

mroeschke commented on pandas/tests/extension/base/methods.py:
```diff
@@ -331,7 +332,8 @@ def test_fillna_length_mismatch(self, data_missing):
 
     # Subclasses can override if we expect e.g Sparse[bool], boolean, pyarrow[bool]
-    _combine_le_expected_dtype: Dtype = np.dtype(bool)
+    # _combine_le_expected_dtype: Dtype = np.dtype(bool)
```
```suggestion
```

MarcoGorelli committed:
Update pandas/tests/extension/base/methods.py

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

mroeschke commented:
Thanks @parthi-siva and @MarcoGorelli

jorisvandenbossche commented on pandas/tests/series/test_constructors.py:
```diff
@@ -572,7 +572,10 @@ def test_constructor_maskedarray(self):
         data[1] = 1
         result = Series(data, index=index)
         expected = Series([0, 1, 2], index=index, dtype=int)
-        tm.assert_series_equal(result, expected)
+        with pytest.raises(AssertionError, match="Series classes are different"):
+            # TODO should this be raising at all?
+            # https://github.com/pandas-dev/pandas/issues/56131
+            tm.assert_series_equal(result, expected)
```
This is a different issue than reported in https://github.com/pandas-dev/pandas/issues/56131 (there is also no `check_dtype=False` here)

This doesn't involve different data types (it's all numpy dtypes), and the two objects created here seemingly are the same:

```python
data = np.ma.masked_all((3,), dtype=int)
data[0] = 0
data[1] = 1
data[2] = 2
index = ["a", "b", "c"]
result = Series(data, index=index)
expected = Series([0, 1, 2], index=index, dtype=int)
```
```
In [29]: result
Out[29]: 
a    0
b    1
c    2
dtype: int64

In [30]: expected
Out[30]: 
a    0
b    1
c    2
dtype: int64

In [31]: result.dtype == expected.dtype
Out[31]: True
```

But apparently we have a bug in the Series constructor that preserves the masked array as underlying value if it has no masked elements:

```
In [32]: result.values
Out[32]: 
masked_array(data=[0, 1, 2],
             mask=[False, False, False],
       fill_value=999999)

In [33]: expected.values
Out[33]: array([0, 1, 2])
```

That seems like a separate bug we should solve.

parthi-siva commented:
> Looks like another bug was hiding because of the check_exact default 😄 #56136
> 
> For now let's xfail the relevant test

@jorisvandenbossche yeah.. there was another issues because of this
